File: 2021/day_6/main.go
package main

import (
	"adventofcode/adventUtils"
	"fmt"
	"strconv"
	"strings"
)

func parseLine(l []string) []int {
	rawNumbers := strings.Split(l[0], ",")
	res := []int{}
	for _, s := range rawNumbers {
		n, err := strconv.Atoi(s)
		if err != nil {
			panic(err)
		}
		res = append(res, n)
	}
	return res
}

func parseToMap(f []int) map[int]int64 {
	m := map[int]int64{}
	for _, d := range f {
		if _, ok := m[d]; ok {
			m[d] += 1
		} else {
			m[d] = 1
		}
	}
	return m
}

func passDay(fish []int) []int {
	res := []int{}
	extra := 0
	for _, f := range fish {
		if f == 0 {
			res = append(res, 6)
			extra += 1
		} else {
			res = append(res, f-1)
		}
	}
	for i := 0; i < extra; i++ {
		res = append(res, 8)
	}
	return res
}

func passDays(fish []int, days int) []int {
	for i := 0; i < days; i++ {
		fish = passDay(fish)
	}
	return fish
}

func passDayCount(fish map[int]int64) map[int]int64 {
	newMap := map[int]int64{}
	for i := 8; i >= 0; i-- {
		if i == 0 {
			newMap[6] += fish[0]
			newMap[8] = fish[0]
		} else {
			newMap[i-1] = fish[i]
		}
	}
	return newMap
}

func countFishAfterDays(fish []int, days int) int64 {
	mappedFish := parseToMap(fish)
	for i := 0; i < days; i++ {
		mappedFish = passDayCount(mappedFish)
	}
	var res int64 = 0
	for _, c := range mappedFish {
		res += c
	}
	return res
}

func day_6() {
	filename := "input_easy.txt"
	rawLines, err := adventUtils.ReadFileAdvent(filename, 6)
	if err != nil {
		fmt.Errorf("%+v\n", err)
		panic(err)
	}
	init := parseLine(rawLines)

	// Part 1
	after80Days := countFishAfterDays(init, 80)
	fmt.Printf("After 80 days there are %d fish.\n", after80Days)

	// Part 2
	after256Days := countFishAfterDays(init, 256)
	fmt.Printf("After 256 days there are %d fish \n", after256Days)

}

func main() {
	adventUtils.Benchmark(day_6)
}

File: 2021/day_6/main.rs
use std::fs;
use std::time::Instant;

fn read_file(path: &str) -> Vec<String> {
    let parsed = fs::read_to_string(path).expect("Could not read file");
    let split_lines = parsed.lines();
    split_lines.map(|s| s.to_string()).collect()
}

fn parse_lines(lines: &Vec<String>) -> Vec<i64> {
    lines[0]
        .split(",")
        .map(|i| i.parse::<i64>().expect("Could not parse number"))
        .collect()
}

fn parse_to_map(fish: &Vec<i64>) -> [i64; 9] {
    let mut res = [0; 9];
    for f in fish {
        res[(*f) as usize] += 1;
    }
    res
}

fn pass_day(fish_map: &[i64; 9]) -> [i64; 9] {
    let mut new_map = [0; 9];
    for i in (1..9).rev() {
        new_map[i - 1] = fish_map[i];
    }
    new_map[8] = fish_map[0];
    new_map[6] += fish_map[0];
    new_map
}

fn count_after_days(fish_map: &[i64; 9], days: i32) -> i64 {
    let mut after_day = *fish_map;
    for _ in 0..days {
        after_day = pass_day(&after_day);
    }
    let mut res = 0;
    for i in after_day {
        res += i;
    }
    res
}

fn day_6() {
    let filename = "2021/day_6/input_easy.txt";
    let lines = read_file(filename);
    let fishes = parse_lines(&lines);
    let fish_map = parse_to_map(&fishes);

    // Part 1
    let fish_80 = count_after_days(&fish_map, 80);
    println!("After 80 days there are {} fish!", fish_80);

    // Part 2
    let fish_256 = count_after_days(&fish_map, 256);
    println!("After 256 days there are {} fish!", fish_256);
}

fn main() {
    let now = Instant::now();
    day_6();
    println!("Rust took {} microseconds", now.elapsed().as_micros());
}
